### ISSUE TITLE
Ability to search remote image, but via uploading instead of URL.

### ISSUE THREAD

**Issue by frmrgr** [dessant/search-by-image#11]
Some websites use UA/cookie check before sending content, so searching image by url doesn't work. The image can be saved to the disk and then searched, but that is kind of slow. It would be really nice to drag and drop and image from another tab into upload tab of your extension to search it as if it was a local file.

**Comment by dessant** [comment]
I agree, some way of searching for images from private sites will have to be implemented.
I'm not sure about dragging and dropping, these sites also tend to declare images in CSS and other methods, so there is nothing to drag and drop in some cases.

**Comment by dessant** [comment]
Could you share a site with open registration for testing?

**Comment by frmrgr** [comment]
Sure, [here's an example](https://jig-san.me/heap/pic/cookietest/). Say if you need another kind of example.

**Comment by dessant** [comment]
Not the final solution, but in 1.7.0 you can copy images declared in `<img>` elements, and paste them in the upload page, without the need for saving them locally.

**Comment by frmrgr** [comment]
This is good for desktop, but it seems that it is problematic on Firefox mobile. Since there is no context menu option, the extension has to be configured to "select image from browser toolbar" mode, but then there is no button to open "drag or paste" page except by changing addon options first. Or do I miss something?

**Comment by dessant** [comment]
There is no `Copy Image` in the context menu on Firefox for Android, so this workaround is mostly for desktop only. Regarding extension configuration, you don't *have* to hide the toolbar popup, it just allows faster searching if you do so, at the expense of being able to change the search mode as you use the extension.

**Comment by frmrgr** [comment]
You are right about the lack of copy image function, I didn't think about that.
Maybe it just makes sense to add twin options to search engines list, like, Google (with Upload), Tineye (with upload) etc, but at least for me current method works well, thanks!

**Comment by dessant** [comment]
I would have liked to just have an option for this. When enabled, the extension checks if the URL returns an image, and downloads it from the context of the page if not. This would have been a more seamless experience, because you would not have needed to think about which page has private images while using the extension.

Unfortunately that implementation would fail in multiple ways:

* The page may serve a dummy image with text about hotlinking being disallowed
* The image may be downloadable from a local server/intranet, but would not be accessible by search engine servers
* The image may be accessible by the user, but the search engine server may be geo-blocked, or unable to access certain URLs (noticed this with engines from China)

Currently the only reliable solution that comes to mind is to have a new search mode that always uploads the image after it was selected instead of trying to guess if the search engine will be able to process the URL or if it will receive the correct image.

**Comment by frmrgr** [comment]
I am not sure about the auto check. It would help against cookies and referrer, but I don't have any knowledge about how often these are the ways the image is protected.

So, my totally personal choice would be one of the following:
1) To have a separate context menu entry (like "Download and Search Image on..." or something like that) enablable in options. This would allow using search all feature as usual. I don't know, if this entry should be inside the main submenu or top-level.
2) To have a button somewhere, maybe in context menu or a the panel, which would toggle the method of searching per domain.

But then again, it's just what I would do for myself; I have no idea how usable that would be for others.

**Comment by dessant** [comment]
A new search mode has been added (Select & Upload) with no per domain settings for now. The search mode can be set separately for the context menu and toolbar, see the new setting in the options page for the context menu and the search mode selector in the toolbar popup. This feature will be part of the next release.
